File: android-host/app/src/main/java/com/huddle/host/data/UndercoverWordStore.kt
package com.huddle.host.data

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

data class UndercoverWordPair(
    val id: String,
    val civilian: String,
    val undercover: String,
    val used: Boolean
)

object UndercoverWordStore {
    private const val PREFS = "huddle_undercover"
    private const val KEY_PAIRS = "word_pairs"
    private const val KEY_COUNT = "undercover_count"
    private const val KEY_MIN_PLAYERS = "min_players"
    private const val KEY_MAX_PLAYERS = "max_players"
    private const val KEY_SEEDED = "seeded"

    const val ABS_MIN_PLAYERS = 3
    const val ABS_MAX_PLAYERS = 12

    private val seedPairs = listOf(
        "咖啡" to "奶茶",
        "月亮" to "太阳",
        "公交" to "地铁",
        "西瓜" to "哈密瓜",
        "铅笔" to "钢笔",
        "电梯" to "扶梯",
        "篮球" to "足球",
        "饺子" to "包子"
    )

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    fun ensureSeeded(context: Context) {
        val p = prefs(context)
        if (p.getBoolean(KEY_SEEDED, false)) return
        if (p.contains(KEY_PAIRS) && !p.getString(KEY_PAIRS, null).isNullOrBlank()) {
            p.edit().putBoolean(KEY_SEEDED, true).apply()
            return
        }
        val arr = JSONArray()
        for ((civilian, undercover) in seedPairs) {
            arr.put(pairJson(UUID.randomUUID().toString(), civilian, undercover, used = false))
        }
        p.edit()
            .putString(KEY_PAIRS, arr.toString())
            .putBoolean(KEY_SEEDED, true)
            .apply()
    }

    fun list(context: Context): List<UndercoverWordPair> {
        ensureSeeded(context)
        val raw = prefs(context).getString(KEY_PAIRS, "[]") ?: "[]"
        val arr = try {
            JSONArray(raw)
        } catch (_: Exception) {
            JSONArray()
        }
        val out = ArrayList<UndercoverWordPair>(arr.length())
        for (i in 0 until arr.length()) {
            val o = arr.optJSONObject(i) ?: continue
            val civilian = o.optString("civilian").trim()
            val undercover = o.optString("undercover").trim()
            if (civilian.isEmpty() || undercover.isEmpty()) continue
            out.add(
                UndercoverWordPair(
                    id = o.optString("id").ifBlank { UUID.randomUUID().toString() },
                    civilian = civilian,
                    undercover = undercover,
                    used = o.optBoolean("used", false)
                )
            )
        }
        return out
    }

    fun unusedCount(context: Context): Int = list(context).count { !it.used }

    fun undercoverCount(context: Context): Int {
        ensureSeeded(context)
        return prefs(context).getInt(KEY_COUNT, 1).coerceIn(1, 3)
    }

    fun setUndercoverCount(context: Context, count: Int) {
        prefs(context).edit().putInt(KEY_COUNT, count.coerceIn(1, 3)).apply()
    }

    fun minPlayers(context: Context): Int {
        ensureSeeded(context)
        val min = prefs(context).getInt(KEY_MIN_PLAYERS, ABS_MIN_PLAYERS)
            .coerceIn(ABS_MIN_PLAYERS, ABS_MAX_PLAYERS)
        val max = prefs(context).getInt(KEY_MAX_PLAYERS, 8)
            .coerceIn(ABS_MIN_PLAYERS, ABS_MAX_PLAYERS)
        return min.coerceAtMost(max)
    }

    fun maxPlayers(context: Context): Int {
        ensureSeeded(context)
        val min = prefs(context).getInt(KEY_MIN_PLAYERS, ABS_MIN_PLAYERS)
            .coerceIn(ABS_MIN_PLAYERS, ABS_MAX_PLAYERS)
        val max = prefs(context).getInt(KEY_MAX_PLAYERS, 8)
            .coerceIn(ABS_MIN_PLAYERS, ABS_MAX_PLAYERS)
        return max.coerceAtLeast(min)
    }

    fun setMinPlayers(context: Context, value: Int) {
        val max = maxPlayers(context)
        val min = value.coerceIn(ABS_MIN_PLAYERS, max)
        prefs(context).edit().putInt(KEY_MIN_PLAYERS, min).apply()
    }

    fun setMaxPlayers(context: Context, value: Int) {
        val min = minPlayers(context)
        val max = value.coerceIn(min, ABS_MAX_PLAYERS)
        prefs(context).edit().putInt(KEY_MAX_PLAYERS, max).apply()
    }

    fun playersLabel(context: Context): String {
        val min = minPlayers(context)
        val max = maxPlayers(context)
        return if (min == max) "$min 人" else "$min–$max 人"
    }

    fun add(context: Context, civilian: String, undercover: String): UndercoverWordPair? {
        val c = civilian.trim()
        val u = undercover.trim()
        if (c.isEmpty() || u.isEmpty()) return null
        if (c == u) return null
        val pair = UndercoverWordPair(
            id = UUID.randomUUID().toString(),
            civilian = c,
            undercover = u,
            used = false
        )
        val pairs = list(context).toMutableList()
        pairs.add(pair)
        save(context, pairs)
        return pair
    }

    fun remove(context: Context, id: String) {
        val pairs = list(context).filter { it.id != id }
        save(context, pairs)
    }

    fun setUsed(context: Context, id: String, used: Boolean) {
        val pairs = list(context).map {
            if (it.id == id) it.copy(used = used) else it
        }
        save(context, pairs)
    }

    fun reuse(context: Context, id: String) = setUsed(context, id, used = false)

    fun reuseAll(context: Context) {
        val pairs = list(context).map { it.copy(used = false) }
        save(context, pairs)
    }

    /** Pick a random unused pair. Does not mark it used. */
    fun pickUnused(context: Context): UndercoverWordPair? {
        val unused = list(context).filter { !it.used }
        if (unused.isEmpty()) return null
        return unused.random()
    }

    fun markUsed(context: Context, id: String) = setUsed(context, id, used = true)

    private fun save(context: Context, pairs: List<UndercoverWordPair>) {
        val arr = JSONArray()
        for (p in pairs) {
            arr.put(pairJson(p.id, p.civilian, p.undercover, p.used))
        }
        prefs(context).edit()
            .putString(KEY_PAIRS, arr.toString())
            .putBoolean(KEY_SEEDED, true)
            .apply()
    }

    private fun pairJson(id: String, civilian: String, undercover: String, used: Boolean) =
        JSONObject()
            .put("id", id)
            .put("civilian", civilian)
            .put("undercover", undercover)
            .put("used", used)
}

File: harmony-host/entry/src/main/ets/shared/runtime/UndercoverWordStore.ts
import type { KvStorage } from "./storage";

export interface UndercoverWordPair { id: string; civilian: string; undercover: string; used: boolean; }

const SEED_PAIRS: string[][] = [
  ["咖啡", "奶茶"], ["月亮", "太阳"], ["公交", "地铁"], ["西瓜", "哈密瓜"],
  ["铅笔", "钢笔"], ["电梯", "扶梯"], ["篮球", "足球"], ["饺子", "包子"]
];

export class UndercoverWordStore {
  private s: KvStorage;
  constructor(s: KvStorage) { this.s = s; }

  ensureSeeded(): void {
    if (this.s.getBool("uc_seeded", false)) return;
    if (this.s.getString("uc_pairs", "") !== "") { this.s.setBool("uc_seeded", true); return; }
    const arr: UndercoverWordPair[] = SEED_PAIRS.map((p: string[]): UndercoverWordPair => {
      return { id: uuid(), civilian: p[0], undercover: p[1], used: false };
    });
    this.save(arr);
    this.s.setBool("uc_seeded", true);
  }

  list(): UndercoverWordPair[] {
    this.ensureSeeded();
    const raw: string = this.s.getString("uc_pairs", "[]");
    let arr: ESObject[] = [];
    try { arr = JSON.parse(raw) as ESObject[]; } catch (e) { arr = []; }
    const out: UndercoverWordPair[] = [];
    for (const o of arr) {
      const civilian: string = String((o as ESObject).civilian != null ? (o as ESObject).civilian : "").trim();
      const undercover: string = String((o as ESObject).undercover != null ? (o as ESObject).undercover : "").trim();
      if (!civilian || !undercover) continue;
      out.push({ id: String((o as ESObject).id != null ? (o as ESObject).id : uuid()), civilian: civilian, undercover: undercover, used: Boolean((o as ESObject).used) });
    }
    return out;
  }

  undercoverCount(): number { this.ensureSeeded(); return clamp(this.s.getNumber("uc_count", 1), 1, 3); }
  setUndercoverCount(v: number): void { this.s.setNumber("uc_count", clamp(v, 1, 3)); }

  minPlayers(): number {
    this.ensureSeeded();
    const min: number = clamp(this.s.getNumber("uc_min", 3), 3, 12);
    const max: number = clamp(this.s.getNumber("uc_max", 8), 3, 12);
    return Math.min(min, max);
  }
  maxPlayers(): number {
    this.ensureSeeded();
    const min: number = clamp(this.s.getNumber("uc_min", 3), 3, 12);
    const max: number = clamp(this.s.getNumber("uc_max", 8), 3, 12);
    return Math.max(min, max);
  }
  setMinPlayers(v: number): void { this.s.setNumber("uc_min", clamp(v, 3, this.maxPlayers())); }
  setMaxPlayers(v: number): void { this.s.setNumber("uc_max", clamp(v, this.minPlayers(), 12)); }
  playersLabel(): string { const a: number = this.minPlayers(), b: number = this.maxPlayers(); return a === b ? `${a} 人` : `${a}–${b} 人`; }

  pickUnused(): UndercoverWordPair | null {
    const unused: UndercoverWordPair[] = this.list().filter((p: UndercoverWordPair): boolean => !p.used);
    return unused.length > 0 ? unused[Math.floor(Math.random() * unused.length)] : null;
  }
  markUsed(id: string): void {
    this.save(this.list().map((p: UndercoverWordPair): UndercoverWordPair => p.id === id ? { id: p.id, civilian: p.civilian, undercover: p.undercover, used: true } : p));
  }
  reuse(id: string): void {
    this.save(this.list().map((p: UndercoverWordPair): UndercoverWordPair => p.id === id ? { id: p.id, civilian: p.civilian, undercover: p.undercover, used: false } : p));
  }
  reuseAll(): void {
    this.save(this.list().map((p: UndercoverWordPair): UndercoverWordPair => ({ id: p.id, civilian: p.civilian, undercover: p.undercover, used: false })));
  }
  add(civilian: string, undercover: string): UndercoverWordPair | null {
    const c: string = civilian.trim(), u: string = undercover.trim();
    if (!c || !u || c === u) return null;
    const pair: UndercoverWordPair = { id: uuid(), civilian: c, undercover: u, used: false };
    const all: UndercoverWordPair[] = this.list(); all.push(pair); this.save(all);
    return pair;
  }
  remove(id: string): void { this.save(this.list().filter((p: UndercoverWordPair): boolean => p.id !== id)); }

  private save(pairs: UndercoverWordPair[]): void {
    this.s.setString("uc_pairs", JSON.stringify(pairs));
    this.s.setBool("uc_seeded", true);
  }
}

function clamp(v: number, lo: number, hi: number): number { return Math.max(lo, Math.min(hi, v)); }
function uuid(): string {
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c: string): string => {
    const r: number = Math.random() * 16 | 0;
    const v: number = c === "x" ? r : (r & 0x3 | 0x8);
    return v.toString(16);
  });
}
